components/TextFields/src/ContainedInputView/private/MDCContainedInputViewLabelAnimator.m

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCContainedInputViewLabelAnimator.h"
+#import "MaterialMath.h"
+
+@interface MDCContainedInputViewLabelAnimator ()
+@end
+
+@implementation MDCContainedInputViewLabelAnimator
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    self.animationDuration = 0.2;
+  }
+  return self;
+}
+
+- (void)layOutLabel:(nonnull UILabel *)floatingLabel
+                 state:(MDCContainedInputViewLabelState)labelState
+      normalLabelFrame:(CGRect)normalLabelFrame
+    floatingLabelFrame:(CGRect)floatingLabelFrame
+            normalFont:(nonnull UIFont *)normalFont
+          floatingFont:(nonnull UIFont *)floatingFont {
+  UIFont *targetFont = normalFont;
+  CGRect targetFrame = normalLabelFrame;
+  BOOL labelShouldHide = NO;
+  switch (labelState) {
+    case MDCContainedInputViewLabelStateFloating:
+      targetFont = floatingFont;
+      targetFrame = floatingLabelFrame;
+      break;
+    case MDCContainedInputViewLabelStateNormal:
+      break;
+    case MDCContainedInputViewLabelStateNone:
+      labelShouldHide = YES;
+      break;
+    default:
+      break;
+  }
+
+  floatingLabel.hidden = labelShouldHide;
+
+  CGRect currentFrame = floatingLabel.frame;
+  CGAffineTransform trasformNeededToMakeViewWithTargetFrameLookLikeItHasCurrentFrame =
+      [self transformFromRect:targetFrame toRect:currentFrame];
+  CATransform3D transformFromValueTransform3D = CATransform3DMakeAffineTransform(
+      trasformNeededToMakeViewWithTargetFrameLookLikeItHasCurrentFrame);
+  CATransform3D transformToValueTransform3D = CATransform3DIdentity;
+
+  floatingLabel.frame = targetFrame;
+  floatingLabel.font = targetFont;
+  floatingLabel.transform = CGAffineTransformIdentity;
```

Comment:
Can this be animated using UIView animation APIs? You'll get additive animations for free that way.

Reply:
I originally tried something like this: 
```objc
  if (self.labelState == MDCContainedInputViewLabelStateFloating) {
    [UIView animateWithDuration:0.15 animations:^{
      self.label.font = self.floatingFont;
      self.label.frame = self.layout.labelFrameFloating;
    }];
  } else {
    [UIView animateWithDuration:0.15 animations:^{
      self.label.font = self.normalFont;
      self.label.frame = self.layout.labelFrameNormal;
    }];
  }
```

But found that it looked like this:
![label_bad](https://user-images.githubusercontent.com/8020010/64178873-3685c800-ce30-11e9-92e3-d7d60d46ec20.gif)

So I got a little fancier with it and used core animation and stuff.

Reply:
It looks like you're only animating the transform with Core Animation. Did you try animating the transform with UIView animateWithDuration?

Reply:
I actually don't think I ever circled back to trying that haha, but I just tried it, and it does seem to work just as well... See 5012e4f

Reply:
`MDCFloatingButton (Animation)` may have some code you can steal.

Reply:
Haha may not need to steal from floating button...